docs/tutorial/programming.md

```diff
@@ -1,10 +1,295 @@
+// 0x3c9 - base pointer for input O (RW)
+// 0x3ca - send configurations to streamer (RW)
+// 0x3cb - performance counter of streamer (RO)
+//------------------------------
+write_csr(0x3c0, LOOP_ITER);
+write_csr(0x3c1, 32);
+write_csr(0x3c2, 32);
+write_csr(0x3c3, 64);
+write_csr(0x3c4, 8);
+write_csr(0x3c5, 8);
+write_csr(0x3c6, 8);
+write_csr(0x3c7, (uint64_t)local_a);
+write_csr(0x3c8, (uint64_t)local_b);
+write_csr(0x3c9, (uint64_t)local_o);
+write_csr(0x3ca, 1);
+```
+
+The sequence above first configures the streamers and then starts the streamers. Recall that our streamer supports the following loop for accessing data:
+
+```
+for(i = 0; i < N; i++):
+  # Parfor equivalent
+  parfor(j = 0; j < 4; j++):
+    target_address[j] = base_address + temporal_stride*i + spatial_stride*j;
+```
+Where `N` is the configured temporal loop-bound register. The `base_address` for each port `A`, `B`, and `OUT` are configured with `(uint64_t)local_a`, `(uint64_t)local_b`, and `(uint64_t)local_o`, respectively.
+
+The temporal strides for each input are 32 bytes, but the output is 64 bytes. These addresses are in bytes. We need 32 for the inputs because that's 8 bytes (64 bits) per PE and we have 4 PEs and therefore, 32 bytes of temporal stride. The output is 64 because we need 16 bytes (128 bits) per PE and we have 4 PEs and therefore, 64 bytes of temporal stride.
+
+All spatial strides are 8 because, for each TCDM port connected to the streamer, it will automatically increment or point to the next bank which is 8 in byte addresses (each bank is 64bit or 8 bytes).
+
+Finally, the start signal for the streamer is configured by writing 1 to the LSB of the start register.
+
+The accelerator registers are configured but take note of the starting address. The code snippet below shows the simple CSR setup. 
+
+```C
+//------------------------------
+// 2nd set the CSRs of the accelerator
+// 0x3cc - mode of the ALU (RW)
+//       - 0 for add, 1 for sub, 2 for mul, 3 for XOR
+// 0x3cd - length of data (RW)
+// 0x3ce - send configurations to accelerator (RW)
+// 0x3cf - busy status (RO)
+// 0x3d0 - performance counter (RO)
+//------------------------------
+write_csr(0x3cc, MODE);
+write_csr(0x3cd, LOOP_ITER);
+write_csr(0x3ce, 1);
+```
+The `MODE` and `LOOP_ITER` are found in the `data.h`. The `LOOP_ITER` is the effective number of loops while considering the spatial parallelism. For example, if we have 100 elements (each is 64-bit wide), then with a spatial parallelism of 4 PEs, we only need to process 25 cycles so the `LOOP_ITER` is 25. 
+
+The last register written on address `0x3ce` is the start signal of the accelerator. Once started, the accelerator and streamer work together to process the data. We can monitor and poll the busy status of the accelerator by reading on register address `0x3cf`. The busy signal is high when the accelerator is still processing data.
+
+```C
+// Do this to poll the accelerator
+while (read_csr(0x3cf)) {
+};
```

Comment:
We need to change this to use the interrupts please

Reply:
Separate PR 😆